include/circt/Dialect/FIRRTL/FIRRTLCanonicalization.td

```diff
@@ -86,6 +86,8 @@ def DropWrite : NativeCodeCall<"dropWrite($_builder, $0, $1)">;
 
 def MoveNameHint : NativeCodeCall<"moveNameHint($0, $1)">;
 
+def SubFieldOrIndex : NativeCodeCall<"subfieldOrIndex($_builder, $_loc, $0, $1)">;
```

Comment:
Can the `NativeCodeCall` be avoided with an equivalent declarative rewrite implementation?

Reply:
Well we could by hoisting the entire body of `subfieldOrIndex` (including the TypeSwitch) into TableGen :upside_down_face: .  But we probably want something better....:

If we split the patterns for ref-of-bundle vs ref-of-vector we could do this--the issue is ref.sub is agnostic to the aggregate type (as long as it has field ID's), but we need to figure out which of subfield or subindex to create.

Let's see if we still want the ref.sub canonicalizers, per your later point.

Reply:
Good call, having two patterns was easy and now it almost entirely lives in the declarative land (other than the type constraints/predicates).